File: leetcode/2406-Divide-Intervals-Into-Minimum-Number-of-Groups/java/Solution.java
public class Solution {
    public int minGroup(int[][] intervals) {
        if (intervals == null || intervals.length == 0) return 0;

        List<int[]> sortedEvents = new ArrayList<>();
        for (int[] itv : intervals) {
            sortedEvents.add(new int[]{itv[0], 1});
            sortedEvents.add(new int[]{itv[1], -1});
        }
        Collections.sort(sortedEvents, new Comparator<int[]>() {
            public int compare(int[] o1, int[] o2) {
                if (o1[0] == o2[0])
                    return o2[1] - o1[1];
                else
                    return o1[0] - o2[0];
            }
        });
        int groups = 0;
        int minGroups = 0;
        for (int[] e : sortedEvents) {
            groups += e[1];
            minGroups = Math.max(minGroups, groups);
        }
        return minGroups;

    }
}

File: leetcode/2406-Divide-Intervals-Into-Minimum-Number-of-Groups/kotlin/Solution.kt
class Solution {
    fun minGroups(intervals: Array<IntArray>): Int {
        if (intervals == null) return 0
        return intervals
            .flatMap { (s, e) -> listOf(Pair(s, 1), Pair(e, -1))}
            .sortedWith( compareBy( {it.first}, {-it.second} ))
            .fold(Pair(0, 0)) { a, e -> Pair(max(a.first, a.second), a.second + e.second)}
            .first
    }
}